### PR TITLE
Avoid unintended copy

### PR DESCRIPTION
**Pull request by mrambausek**
Mesh, Topology and Geometry have copy construction but no copy assignment. I propose to also "delete" the copy constructors, which are probably anyway not supposed to be called. Deleting them revealed possible unintended copies. The first commit of this PR deals with the unintended copies, the second with the deletion of constructors. I chose that order because I do not know whether copy construction is totally unwanted. Note however, that I did not find any intended copy in the code.

### CODE REVIEW

**Comment by mrambausek** [comment]
Okay, the python tests reveal that copy construction is currently required because Mesh stores Geometry and Topology as object not as shared_ptr, but python handles all of them as shared_ptr. Shall these copies be avoided as well, ie. store shared_ptr<(const) Topology|Geometry>?

**Comment by mrambausek** [comment]
If the copies at the python level should stay, second commit just might be dropped.

File: cpp/dolfinx/mesh/Geometry.h [hunk 1]
```diff
@@ -64,7 +64,7 @@ class Geometry
   }
 
   /// Copy constructor
```
Comment (garth-wells) [inline cpp/dolfinx/mesh/Geometry.h:66]: Probably want to change `///` -> `//` so the deleted constructor doesn't get documented. Not sure what doxygen does with deleted constructors.

File: cpp/dolfinx/mesh/Mesh.h [hunk 2]
```diff
@@ -90,7 +90,7 @@ class Mesh
 
   /// Copy constructor
```
Comment (garth-wells) [inline cpp/dolfinx/mesh/Mesh.h:91]: Probably want to change `///` -> `//`

File: cpp/dolfinx/mesh/Mesh.h [hunk 3]
```diff
@@ -90,7 +90,7 @@ class Mesh
...
   /// @param[in] mesh Mesh to be copied
```
Comment (garth-wells) [inline cpp/dolfinx/mesh/Mesh.h:92]: Deleted

File: cpp/dolfinx/mesh/Topology.h [hunk 4]
```diff
@@ -67,7 +67,7 @@ class Topology
   }
 
   /// Copy constructor
```
Comment (garth-wells) [inline cpp/dolfinx/mesh/Topology.h:69]: Probably want to change `///` -> `//`

**Comment by mrambausek** [comment]
Actually, I hesitate because the python version of Mesh's constructor using Topology and Geometry either requires copies or Mesh has to hold these members using shared_ptr<Topology|Geometry>
The copy in this case probably probably does not match the usual shared ownership of python.
What do you prefer? In any case I can just go ahead with adding a an explicit copy members.

**Comment by garth-wells** [comment]
Yes, I see the issue now in the Python interface so let's keep the copy constructor and assignment operator.

**Comment by mrambausek** [comment]
Just pushed a proposal with explicit copy and copy assignment. The Mesh constructor in python is probably the only rare case where this shows up. But feel free to drop this and I will just create a small PR with only the first commit.

**Comment by garth-wells** [comment]
Having considered the use cases with the new `Mesh` design, I'm no longer so sure it's best to hide the copy constructor. There are some common cases where a copy is expected.

@mrambausek  Could you make a PR that avoids the unnecessary copy so we can discuss the copy semantics separately from the bug fix?

**Comment by mrambausek** [comment]
See  #981

**Pull request closed without merging**